### ISSUE TITLE
[autoscaler][gcp] NodeProvider: Terminate nodes in parallel

### ISSUE THREAD

**Issue by yoavhacohen** [ray-project/ray#26239]
### Description

The current implementation of `terminate_nodes` terminates nodes one by one:
```python
    def terminate_nodes(self, node_ids: List[str]) -> Optional[Dict[str, Any]]:
        """Terminates a set of nodes.

        May be overridden with a batch method, which optionally may return a
        mapping from deleted node ids to node metadata.
        """
        for node_id in node_ids:
            logger.info("NodeProvider: {}: Terminating node".format(node_id))
            self.terminate_node(node_id)
        return None
```

Hence, `ray down` takes a lot of time when there are many nodes.

Instead, the loop should be parallelized, so all nodes will be terminated together.

### Use case

When using many nodes it takes a lot of time to `ray down` to complete.

**Comment by DmitriGekhtman** [comment]
An ideal fix would be to make the `terminate_node` call non-blocking for the NodeProvider implementation you are using.

Alternatively, the NodeProvider implementation you are using could implement an override which batches the calls, as suggested in the doc string.

Which implementation are you using? We could consider optimizing that particular NodeProvider.

**Comment by yoavhacohen** [comment]
Thanks, I'm using GCP.
An optimized version of it would be really helpful!

**Comment by bradhilton** [comment]
I'm also running into this issue, it would be nice if the GCP NodeProvider supported bulk termination. 🙏

**Comment by DmitriGekhtman** [comment]
Could also use a thread pool and the terminate_node (note the singular "node") method in the ray down code.
Should be straightforward to do... feel to open a PR if you're reading this.

**Comment by wuisawesome** [comment]
Adding some more tags, using the batch api sounds like a good idea. Would encourage anyone with bandwidth to contribute it if they can though.

**Comment by Dan-Yeh** [comment]
Hi, can I work on this?

**Comment by wuisawesome** [comment]
Let's say this issue was scoped to just GCP since no one has mentioned this as problematic on AWS, and consider it closed by https://github.com/ray-project/ray/pull/34455. 

If anyone notices a separate issue on other cloud provides let's deal with track that as a separate issue.

**Comment by rickyyx** [comment]
The change was reverted since it had issues with race - we should be using  approaches outline in https://github.com/googleapis/google-api-python-client/blob/main/docs/thread_safety.md to implement concurrent deletion of instances requests IMO.

**Comment by mimiliaogo** [comment]
@rickyyx Not sure if you've already opened a PR to solve this, if not, can I take this?

**Comment by rickyyx** [comment]
Hey sure @mimiliaogo - feel free to take a look at this.

**Comment by mimiliaogo** [comment]
@rickyyx For the thread saftey issue, I'm thinking if we can simply modify these two lines
https://github.com/ray-project/ray/blob/5e4b1bc58807359a9400c6333d9d79923364fb8a/python/ray/autoscaler/_private/gcp/node.py#L583
https://github.com/ray-project/ray/blob/5e4b1bc58807359a9400c6333d9d79923364fb8a/python/ray/autoscaler/_private/gcp/node.py#L601
to
```
 .execute(http=self.get_new_authorized_http(self.resource._http))
 ```
 So that each request will have its own instance of httplib2.Http().
 
 And then we can add back the original implmentation of parallel termination.

**Comment by mimiliaogo** [comment]
@rickyyx However, it appears that the original implementation of parallel termination allows only one thread to execute the termination code at a time due to the lock here. I doubt that this can still achieve concurrency.
https://github.com/Dan-Yeh/ray/blob/51969a9296237594cc2e71a3426d8b85d81031e6/python/ray/autoscaler/_private/gcp/node_provider.py#L204

IMO, we'll need to implement a more fine-grained lock where we need global lock for `_get_cached_node()` and a per-node lock for google API `resource.stop_instance(node_id=node_id)` or `resource.delete_instance(node_id=node_id)`.
Even probably we don't need a per-node lock, as Google might have already implemented this.

Since this is my first issue here, I have a lot of uncertainties. Please let me know if you have any suggestions! Thanks!

**Comment by nh-atuan** [comment]
I'd like to take this issue and work on a fix. I'll investigate the current GCP NodeProvider implementation and submit a PR if no one is actively working on it.
